### PR TITLE
An experimental 'Show Extraction' command (grant feature wish #4129)

### LINKED ISSUE #4129: Extraction window: show interactively an extracted definition during its proof
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#4129
From: @jonleivent
Reported version: 8.5
CC:   coq-bugs-redist@lists.gforge.inria.fr

### PR DESCRIPTION
**Pull request by letouzey**
Attempt to extract the current ongoing proof (request by Clément Pit-Claudel @cpitclaudel on coqdev, and also #4129). Evars are handled as axioms.

This PR is based on top of PR #6925 .

**Kind:** feature (grant feature wish)

Fixes / closes #4129

- [ ] Corresponding documentation was added / updated.
- [X] Entry added in CHANGES.

### CODE REVIEW

**Comment by maximedenes** [comment]
Seems great! Can you add some tests, though?

**Comment by letouzey** [comment]
I've added a minimal test script. Since this `Show Extraction` is a display command, in theory this test should rather be in `test-suite/output`. But these output tests are just a pain, in particular for something as pretty-print-sensitive as the extraction. At least we'll know when the commands fail. Let me also stress that this new command is *experimental*, and will probably fail for more advanced usages.

**Comment by maximedenes** [comment]
I'm not asking to do it now, but what would be the right way to test such features?

File: plugins/extraction/extract_env.ml [hunk 1]
```diff
@@ -135,19 +135,22 @@ let check_arity env cb =
...
+          | Fix ((_,j),recd) when Int.equal i j -> check_arity env cb; (true,recd)
 	  | CoFix (j,recd) when Int.equal i j -> check_arity env cb; (false,recd)
 	  | _ -> raise Impossible)
     | Undef _ | OpaqueDef _ -> raise Impossible
 
 let prec_declaration_equal (na1, ca1, ta1) (na2, ca2, ta2) =
   Array.equal Name.equal na1 na2 &&
-  Array.equal Constr.equal ca1 ca2 &&
-  Array.equal Constr.equal ta1 ta2
+  Array.equal (EConstr.eq_constr Evd.empty) ca1 ca2 &&
```
Comment (gares) [inline plugins/extraction/extract_env.ml]: `Evd.empty` should not even exist, please use `Evd.from_env env` where env is the global one if you don't want to thread it around.  The point is that `Evd` today also contains universes as in *universe name* and *universe constraints*. I think you probably don't care about the latter, but some code may `assert false` if it faces a universe name not part of the evar map.

Why `Evd.empty` has not *big fat warning* attached to is a mystery to me CC @ppedrot

**Review by gares — CHANGES_REQUESTED** [review]
I'm commenting just the Evd.empty thing, I'm not competent on the rest of the patch

File: plugins/extraction/extract_env.ml (hunk 1 above)
Comment (ejgallego) [reply plugins/extraction/extract_env.ml]: > Why Evd.empty has not big fat warning attached to is a mystery to me CC @ppedrot

Yes! Can we put this in 8.8? In fact as it is used in so many parts of the codebase we tend to think it is OK.

**Comment by letouzey** [comment]
Replaced all remaining `Evd.empty` by `Evd.from_env` as suggested by @gares.

**Review by gares — APPROVED** [review]

**Comment by Zimmi48** [comment]
I would have liked some documentation for this feature even if it's experimental. The entry in CHANGES is not enough.

**Comment by letouzey** [comment]
Ok for a short doc. Btw, what's the status of the doc migration to Sphinx ? Will it be the official 8.8 doc ?

**Comment by maximedenes** [comment]
> Btw, what's the status of the doc migration to Sphinx ? Will it be the official 8.8 doc ?

Yes. The infrastructure PR is going to be integrated, and the chapters moved one by one from the old manual.

**Comment by Zimmi48** [comment]
@letouzey if you want to wait until the extraction chapter porting has been integrated, that's fine as well. In that case, just open a new issue as a reminder, self-assign it and put the 8.8.0 milestone on it.

**Pull request merged**